src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp

```diff
@@ -0,0 +1,327 @@
+/*
+  Copyright (C) 2020 Equinor ASA
+  Copyright (C) 2020 SINTEF Digital
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  ?You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <opm/parser/eclipse/Parser/ParserKeywords/A.hpp>
+
+#include <opm/parser/eclipse/EclipseState/NumericalAquifer.hpp>
+#include <opm/parser/eclipse/EclipseState/Aqucon.hpp>
+#include <opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp>
+#include <opm/parser/eclipse/EclipseState/Grid/EclipseGrid.hpp>
+#include <opm/parser/eclipse/EclipseState/Grid/TranCalculator.hpp>
+#include <opm/parser/eclipse/EclipseState/Grid/NNC.hpp>
+#include <opm/parser/eclipse/Deck/Deck.hpp>
+#include <opm/parser/eclipse/Deck/DeckRecord.hpp>
+#include <opm/common/OpmLog/OpmLog.hpp>
+
+#include <iostream>
+
+#include "AquiferHelpers.hpp"
+
+namespace Opm {
+
+NumericalAquifers::NumericalAquifers(const Deck& deck, const EclipseGrid& grid, const FieldPropsManager& field_props)
+{
+    using AQUNUM=ParserKeywords::AQUNUM;
+    if ( !deck.hasKeyword<AQUNUM>() ) return;
+
+    // there might be multiple keywords of keyword AQUNUM, it is not totally
+    // clear about the rules here. For now, we take care of all the keywords
+    const auto& aqunum_keywords = deck.getKeywordList<AQUNUM>();
+    for (const auto& keyword : aqunum_keywords) {
+        for (const auto& record : *keyword) {
+            const NumericalAquiferCell aqu_cell(record, grid, field_props);
+            // TODO: we should check whether the grid cell is active or NOT
+            // NOT sure what to do if the cell is inactive
+            if (this->hasCell(aqu_cell.global_index)) {
+                std::ostringstream ss;
+                ss << "NUMERICAL AQUIFER CELL AT GRID CELL { " << aqu_cell.I + 1
+                   << " " << aqu_cell.J + 1 << " " << aqu_cell.K + 1 << " } IS DECLARED MORE THAN ONCE";
+                throw std::logic_error(ss.str());
+            } else {
+                this->addAquiferCell(aqu_cell);
+                auto& cells = this->aquifer_cells_;
```

Comment:
Would it not be natural to add the `aqufer_cells_` manipulation to the `addAquiferCell()` function?